[SIL] Trivial storage locations remain valid after consumption.

nate-chandler opened the pull request:
Remove some overzealous verification:

Previously, memory locations whose stored type was trivial were treated just like those whose type was not.

Here, this is loosened so that once a location is initialized, it remains valid despite "consuming uses" unless and until it is _deallocated_.

nate-chandler committed:
[SILVerifier] Loosen checks on trivial memory.

Legalize passing "consuming" uses of addresses containing trivial types.
Such types are not destroyed and their destruction is a no-op anyway.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please clean test linux platform

nate-chandler commented on include/swift/SIL/MemoryLocations.h:
```diff
@@ -247,22 +255,38 @@ class MemoryLocations {
+        if (!preserveTrivialLocations || evenTrivial) {
+          return loc->subLocations;
+        } else {
+          return loc->subLocations & getNonTrivialLocations();
```
Maybe it would be better to push this into bitwise anding into the MemoryVerifier instead of adding this mode to MemoryLocations..

eeckstein replied:
I think it doesn't matter. MemoryLocations are only used in the MemoryLifetimeVerifier and DestroyHoisting. DestroyHoisting is not enabled anymore and we can delete it. So The MemoryLifetimeVerifier will be the only user of MemoryLocations.

nate-chandler replied:
Removed this new flag `preserveTrivialLocations`, the old flag `handleTrivialLocations` and `DestroyHoisting`.

eeckstein commented on include/swift/SIL/MemoryLocations.h:
```diff
@@ -201,10 +201,18 @@ class MemoryLocations {
   /// If true, also analyze trivial memory locations.
   bool handleTrivialLocations;
 
+  /// If true, trivial values aren't cleared or killed unless "evenTrivial=true"
```
Actually, there is no need to make this configurable. It can be the default. Given that DestroyHoisting is not used anymore and can be deleted.

nate-chandler replied:
Removed the new flag and the preexisting flag.

nate-chandler committed:
[MemoryLocations] Removed unused flag.

It was used by the old DestroyHoisting pass which has been removed.

nate-chandler committed:
[MemoryVerifier] Trivial locs aren't destroyed.

Once valid, only invalidate storage for a trivial value by way of a
dealloc_stack.  A consuming use of the location does not destroy it.

nate-chandler commented:
@swift-ci please test

eeckstein reviewed (approved):
thanks, lgtm!

You'll probably need to remove the `SILOptimizer/destroy_hoisting.sil` test as well

nate-chandler commented:
@swift-ci please test

atrick reviewed (approved):
This is one valid model of initialized memory at the SIL level. Another valid model would allow us to reason about uninitialized memory independent of the type in memory. The second model would restrict the space of valid SIL, removing the burden on every SIL analysis and utility to handle address operations differently in the presence of trivial types.
Is there a motivation for using the less restrictive model? Is there a source-level construct that requires this? Or is there a particular optimization that we want?

nate-chandler commented:
@swift-ci please test
